File: January 23/25 Jan/2359. Find Closest Node to Given Two Nodes LC.cpp
class Solution {
public:
    int closestMeetingNode(vector<int>& edges, int node1, int node2) {
        int n=edges.size(), k=0, qsize;
        vector<int> nodes={node1,node2};
        
        vector<vector<int>> distance(2,vector<int>(n,INT_MAX));

        for(int i=0;i<2;i++)
        {
            int node = nodes[i];
            queue<int> q;
            vector<bool> visited(n,false);
            q.push(node);
            k=0;
            
            while(!q.empty())
            {
                qsize = q.size();
                for(int j=0;j<qsize;j++)
                {
                    int next = q.front();
                    q.pop();
                    if(visited[next]==false)
                    {
                        if(edges[next]!=-1 )
                            q.push(edges[next]);
                        distance[i][next]=k;
                        visited[next]=true;
                        //cout<<edges[next]<<" "<<k<<endl;
                    }
                }
                k++;
            }

            cout<<endl;
        }
        
        int closest=-1;
        int minDistance=INT_MAX;

        for(int i=0;i<n;i++)
        {
            //cout<<i<<" : "<<distance[0][i]<<" , "<<distance[1][i]<<endl;
            if(minDistance > max(distance[0][i],distance[1][i]))
            {
                closest=i;
                minDistance = max(distance[0][i],distance[1][i]);
            }
        }

        return closest;
    }
};

File: MAY 2025/30 May/2359. Find Closest Node to Given Two Nodes LC.java
class Solution {
    public int[] findDistance(int[] edges, int src)
    {
        int N = edges.length;
        int[] distance = new int[N];
        Arrays.fill(distance, -1);
        distance[src] = 0;
        int node = src, dist=0;

        while(true)
        {
            node = edges[node];
            System.out.println(node);
            if(node==-1 || distance[node]!=-1)
                break;
            distance[node] = ++dist;
        }
        return distance;
    }

    public int closestMeetingNode(int[] edges, int node1, int node2) {
        int[] distance1 = findDistance(edges, node1);
        int[] distance2 = findDistance(edges, node2);

        int N = edges.length;
        int meetingDistance = N+1, meetingNode = N+1;
        for(int i=0; i<N; i++)
        {
            if(distance1[i]==-1 || distance2[i]==-1)
                continue;
            
            // System.out.println(i+" "+distance1[i]+" "+distance2[i]);

            if(Math.max(distance1[i], distance2[i])<meetingDistance)
            {
                meetingDistance = Math.max(distance1[i], distance2[i]);
                meetingNode = i;
            }
            else if(Math.max(distance1[i], distance2[i])==meetingDistance)
            {
                meetingDistance = Math.max(distance1[i], distance2[i]);
                meetingNode = Math.min(meetingNode, i);
            }
        }
        return ( meetingNode==(N+1) ? -1 : meetingNode);
    }
}
